src/hotspot/cpu/s390/gc/shared/barrierSetAssembler_s390.cpp

```diff
@@ -68,6 +72,31 @@ void BarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet decorators,
   }
 }
 
+void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm, Register tmp) {
+  BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+  if (bs_nm == nullptr) {
+    return;
+  }
+
+  assert_different_registers(tmp, Z_R1_scratch);
+
+  __ block_comment("nmethod_entry_barrier (nmethod_entry_barrier) {");
+
+    // Load jump addr:
+    __ load_const(Z_R1_scratch, (uint64_t)StubRoutines::zarch::nmethod_entry_barrier()); // 2*6 bytes
+
+    // Load value from current java object:
+    __ z_lg(tmp, in_bytes(bs_nm->thread_disarmed_offset()), Z_thread); // 6 bytes
+
+    // Compare to current patched value:
+    __ z_cfi(tmp, /* to be patched */ -1); // 6 bytes (2 + 4 byte imm val)
+
```

Comment:
What about the value at `thread_disarmed_offset()`? Is it 4 bytes or 8 bytes? If it is 4 bytes, you should load it with
`z_l(tmp, ...);` or 
`z_ly(tmp, ...);` if the offset might be negative.
Similar size considerations on the compare:
`z_cfi(tmp, ...);` is good for 4-byte value.
`z_cgfi(tmp, ...);` is good for 8-byte values. Caution: signed compare. Sign-extension of the immediate may produce unexpected behaviour.

Reply:
One more thing:
What's the value range of thread_disarmed? Does it use less than 16 bits in all cases? Then you could exploit a storage-immediate variant of compare if the offset is positive:
z_clfhsi(in_bytes(bs_nm->thread_disarmed_offset()), Z_thread, /* to be patched */ -1);
z_clghsi(in_bytes(bs_nm->thread_disarmed_offset()), Z_thread, /* to be patched */ -1); 
2-byte immediate in both cases. Caution: sign extension!
No need for a tmp register anymore!

Reply:
`_nmethod_disarm_value` is 64 bit of which the high order 32 bits are only used on aarch64. Other platforms use a 4 Byte access, so `z_cfi` is correct.

Reply:
Yes, `_nmethod_disarm_value = (uint64_t)(uint32_t)value;`